File: backend/app/schemas/additional_services.py
from __future__ import annotations

from datetime import datetime, date
from decimal import Decimal
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, computed_field, model_validator

ServiceUnitLiteral = Literal["piece", "person", "hour", "package"]
ServiceOrderStatusLiteral = Literal[
    "draft",
    "confirmed",
    "in_progress",
    "completed",
    "cancelled",
    "on_hold",
]
ServiceItemStatusLiteral = Literal[
    "pending",
    "scheduled",
    "in_progress",
    "delivered",
    "cancelled",
    "quoted",
]
ServiceScheduleStatusLiteral = Literal[
    "reserved",
    "confirmed",
    "completed",
    "no_show",
    "cancelled",
]


class ServiceBase(BaseModel):
    code: str = Field(..., min_length=1, max_length=64)
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    category: Optional[str] = None
    unit: ServiceUnitLiteral = "piece"
    base_price: Decimal = Decimal("0")
    estimated_cost: Decimal = Decimal("0")
    cost_currency: str = Field("PLN", min_length=3, max_length=3)
    currency: str = Field("PLN", min_length=3, max_length=3)
    vat_rate: Decimal = Decimal("23")
    requires_schedule: bool = False
    requires_candidate: bool = False
    result_document_type: Optional[str] = None
    requires_documents: Optional[List[str]] = None
    sla_hours: Optional[int] = None
    is_active: bool = True
    meta: Dict[str, Any] = Field(default_factory=dict)


class ServiceCreate(ServiceBase):
    pass


class ServiceUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    category: Optional[str] = None
    unit: Optional[ServiceUnitLiteral] = None
    base_price: Optional[Decimal] = None
    estimated_cost: Optional[Decimal] = None
    cost_currency: Optional[str] = Field(None, min_length=3, max_length=3)
    currency: Optional[str] = Field(None, min_length=3, max_length=3)
    vat_rate: Optional[Decimal] = None
    requires_schedule: Optional[bool] = None
    requires_candidate: Optional[bool] = None
    result_document_type: Optional[str] = None
    requires_documents: Optional[List[str]] = None
    sla_hours: Optional[int] = None
    is_active: Optional[bool] = None
    meta: Optional[Dict[str, Any]] = None


class ServiceOut(ServiceBase):
    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: lambda v: float(v)},
    )

    id: str
    tenant_id: str
    created_at: datetime
    updated_at: datetime
    metrics_orders_count: int = Field(
        0,
        description="Distinct non-cancelled orders that include this catalog service (non-cancelled lines).",
    )
    metrics_revenue_completed: Decimal = Field(
        Decimal("0"),
        description="Sum of line amounts on completed orders (excludes cancelled lines). Mixed currencies are summed numerically.",
    )


class ServiceAttachmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    item_id: str
    file_id: str
    label: Optional[str] = None
    created_at: datetime


class ServiceScheduleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    item_id: str
    provider: Optional[str] = None
    slot_start: Optional[datetime] = None
    slot_end: Optional[datetime] = None
    location: Optional[str] = None
    status: ServiceScheduleStatusLiteral
    meta: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime


class ServiceItemOut(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: lambda v: float(v)},
    )

    id: str
    tenant_id: str
    order_id: str
    service_id: str
    qty: Decimal
    unit_price: Decimal
    estimated_cost: Decimal
    actual_cost: Optional[Decimal] = None
    cost_currency: str
    cost_source: Optional[str] = None
    cost_status: str
    vat_rate: Decimal
    amount: Decimal
    status: ServiceItemStatusLiteral
    required_documents: Optional[List[str]] = None
    result_document_type: Optional[str] = None
    meta: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    service: Optional[ServiceOut] = None
    schedules: List[ServiceScheduleOut] = Field(default_factory=list)
    attachments: List[ServiceAttachmentOut] = Field(default_factory=list)


class ServiceItemCreate(BaseModel):
    service_id: Optional[str] = None
    service_code: Optional[str] = None
    qty: Decimal = Decimal("1")
    unit_price: Optional[Decimal] = None
    estimated_cost: Optional[Decimal] = None
    actual_cost: Optional[Decimal] = None
    cost_currency: Optional[str] = Field(None, min_length=3, max_length=3)
    cost_source: Optional[str] = None
    cost_status: Optional[str] = None
    vat_rate: Optional[Decimal] = None
    required_documents: Optional[List[str]] = None
    result_document_type: Optional[str] = None
    meta: Optional[Dict[str, Any]] = None


class ServiceOrderBase(BaseModel):
    candidate_id: Optional[str] = None
    vacancy_id: Optional[str] = None
    company_id: Optional[str] = None
    currency: str = Field("PLN", min_length=3, max_length=3)
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    notes: Optional[str] = None
    assigned_to: Optional[str] = None
    audit: Optional[Dict[str, Any]] = None


class ServiceOrderCreate(ServiceOrderBase):
    requested_by: Optional[str] = None
    items: List[ServiceItemCreate] = Field(default_factory=list)


class ServiceOrderUpdate(BaseModel):
    # Accepts canonical statuses plus legacy aliases (quoted/approved/scheduled/delivered/refunded).
    status: Optional[str] = None
    notes: Optional[str] = None
    assigned_to: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    audit: Optional[Dict[str, Any]] = None


class ServiceOrderOut(ServiceOrderBase):
    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: lambda v: float(v)},
    )

    own_company_id: Optional[str] = None
    id: str
    tenant_id: str
    status: ServiceOrderStatusLiteral
    total_amount: Decimal
    vat_total: Decimal
    requested_by: str
    created_at: datetime
    updated_at: datetime
    items: List[ServiceItemOut] = Field(default_factory=list)

    @computed_field
    def client_id(self) -> Optional[str]:
        return self.company_id

    @model_validator(mode="before")
    @classmethod
    def _coerce_audit(cls, data: Any) -> Any:
        # DB may contain NULL in audit; API should return a stable object.
        if isinstance(data, dict):
            if data.get("audit") is None:
                data["audit"] = {}
            return data
        try:
            audit = getattr(data, "audit", None)
            if audit is None:
                setattr(data, "audit", {})
        except Exception:
            pass
        return data


class ServiceScheduleCreate(BaseModel):
    provider: Optional[str] = None
    slot_start: Optional[datetime] = None
    slot_end: Optional[datetime] = None
    location: Optional[str] = None
    status: ServiceScheduleStatusLiteral = "reserved"
    meta: Optional[Dict[str, Any]] = None


class ServiceScheduleUpdate(BaseModel):
    provider: Optional[str] = None
    slot_start: Optional[datetime] = None
    slot_end: Optional[datetime] = None
    location: Optional[str] = None
    status: Optional[ServiceScheduleStatusLiteral] = None
    meta: Optional[Dict[str, Any]] = None


class ServiceAttachmentCreate(BaseModel):
    file_id: str
    label: Optional[str] = None


class ResultDocumentPayload(BaseModel):
    document_type: str
    number: Optional[str] = None
    status: str = "approved"
    issued_at: Optional[date] = None
    expires_at: Optional[date] = None
    file_id: Optional[str] = None
    extra: Dict[str, Any] = Field(default_factory=dict)


class ServiceItemDeliverPayload(BaseModel):
    status: Optional[ServiceItemStatusLiteral] = "delivered"
    result_document: Optional[ResultDocumentPayload] = None
    attachments: List[ServiceAttachmentCreate] = Field(default_factory=list)
    meta: Optional[Dict[str, Any]] = None


class ServiceOrderSummary(BaseModel):
    order: ServiceOrderOut
    blocking_items: List[ServiceItemOut] = Field(default_factory=list)
    missing_documents: Dict[str, List[str]] = Field(default_factory=dict)

File: hostflow-frontend/src/api/additionalServices.ts
import { api } from './client'
import type {
  AdditionalService,
  AdditionalServiceAttachment,
  AdditionalServiceItem,
  AdditionalServiceOrder,
  AdditionalServiceOrderSummary,
  AdditionalServiceSchedule,
  ServiceOrderStatus,
  ServiceScheduleStatus,
  ServiceUnit,
} from './types'

export interface CreateAdditionalServiceInput {
  code: string
  name: string
  description?: string | null
  category?: string | null
  unit?: ServiceUnit
  base_price?: number
  estimated_cost?: number
  cost_currency?: string
  currency?: string
  vat_rate?: number
  requires_schedule?: boolean
  requires_candidate?: boolean
  result_document_type?: string | null
  requires_documents?: string[]
  sla_hours?: number | null
  is_active?: boolean
  meta?: Record<string, any>
}

export type UpdateAdditionalServiceInput = Partial<CreateAdditionalServiceInput>

export async function listAdditionalServices(includeInactive = false, includeMetrics = false) {
  const params: Record<string, boolean> = {}
  if (includeInactive) params.include_inactive = true
  if (includeMetrics) params.include_metrics = true
  const { data } = await api.get<AdditionalService[]>(
    '/services',
    Object.keys(params).length ? { params } : undefined
  )
  return data
}

export async function createAdditionalService(payload: CreateAdditionalServiceInput) {
  const { data } = await api.post<AdditionalService>('/services', payload)
  return data
}

export async function updateAdditionalService(id: string, payload: UpdateAdditionalServiceInput) {
  const { data } = await api.patch<AdditionalService>(`/services/${id}`, payload)
  return data
}

export interface CreateServiceOrderItemInput {
  service_id?: string
  service_code?: string
  qty?: number
  unit_price?: number
  estimated_cost?: number
  actual_cost?: number
  cost_currency?: string
  cost_source?: string
  cost_status?: string
  vat_rate?: number
  required_documents?: string[]
  result_document_type?: string | null
  meta?: Record<string, any>
}

export interface CreateServiceOrderInput {
  candidate_id?: string | null
  vacancy_id?: string | null
  company_id?: string | null
  currency?: string
  notes?: string | null
  assigned_to?: string | null
  audit?: Record<string, any>
  items: CreateServiceOrderItemInput[]
}

export interface ServiceOrderQuery {
  candidateId?: string
  vacancyId?: string
  companyId?: string
  status?: ServiceOrderStatus | ServiceOrderStatus[]
}

export async function listServiceOrders(params: ServiceOrderQuery = {}) {
  const query: Record<string, any> = {}
  if (params.candidateId) query.candidate_id = params.candidateId
  if (params.vacancyId) query.vacancy_id = params.vacancyId
  if (params.companyId) query.company_id = params.companyId
  if (params.status) query.status = params.status

  const { data } = await api.get<AdditionalServiceOrder[]>('/service-orders', {
    params: Object.keys(query).length ? query : undefined,
  })
  return data
}

export async function getServiceOrder(orderId: string) {
  const { data } = await api.get<AdditionalServiceOrder>(`/service-orders/${orderId}`)
  return data
}

export async function getServiceOrderSummary(orderId: string) {
  const { data } = await api.get<AdditionalServiceOrderSummary>(
    `/service-orders/${orderId}/summary`
  )
  return data
}

export async function createServiceOrder(payload: CreateServiceOrderInput) {
  const { data } = await api.post<AdditionalServiceOrder>('/service-orders', payload)
  return data
}

export interface UpdateServiceOrderInput {
  status?: ServiceOrderStatus
  notes?: string | null
  assigned_to?: string | null
  audit?: Record<string, any> | null
}

export async function updateServiceOrder(orderId: string, payload: UpdateServiceOrderInput) {
  const { data } = await api.patch<AdditionalServiceOrder>(
    `/service-orders/${orderId}`,
    payload
  )
  return data
}

export async function addServiceOrderItem(
  orderId: string,
  payload: CreateServiceOrderItemInput
) {
  const { data } = await api.post<AdditionalServiceItem>(
    `/service-orders/${orderId}/items`,
    payload
  )
  return data
}

export interface CreateServiceScheduleInput {
  provider?: string | null
  slot_start?: string | null
  slot_end?: string | null
  location?: string | null
  status?: ServiceScheduleStatus
  meta?: Record<string, any>
}

export type UpdateServiceScheduleInput = Partial<CreateServiceScheduleInput>

export async function addServiceSchedule(
  itemId: string,
  payload: CreateServiceScheduleInput
) {
  const { data } = await api.post<AdditionalServiceSchedule>(
    `/service-items/${itemId}/schedule`,
    payload
  )
  return data
}

export async function updateServiceSchedule(
  scheduleId: string,
  payload: UpdateServiceScheduleInput
) {
  const { data } = await api.patch<AdditionalServiceSchedule>(
    `/service-schedule/${scheduleId}`,
    payload
  )
  return data
}

export interface DeliverServiceItemInput {
  status?: 'delivered' | 'in_progress' | 'cancelled'
  result_document?: {
    document_type: string
    status?: string
    issued_at?: string
    expires_at?: string
    number?: string
    file_id?: string
    extra?: Record<string, any>
  }
  attachments?: Array<{
    file_id: string
    label?: string | null
  }>
  meta?: Record<string, any>
}

export async function deliverServiceItem(itemId: string, payload: DeliverServiceItemInput) {
  const { data } = await api.post<AdditionalServiceItem>(
    `/service-items/${itemId}/deliver`,
    payload
  )
  return data
}

export async function addServiceAttachment(
  itemId: string,
  payload: { file_id: string; label?: string | null }
) {
  const { data } = await api.post<AdditionalServiceAttachment>(
    `/service-items/${itemId}/attachments`,
    payload
  )
  return data
}
